server/src/main/java/org/elasticsearch/cluster/routing/allocation/allocator/DesiredBalanceComputer.java

```diff
@@ -0,0 +1,303 @@
+        final var clusterInfoSimulator = new ClusterInfoSimulator(routingAllocation.clusterInfo());
+
+        if (routingNodes.isEmpty()) {
+            return new DesiredBalance(desiredBalanceInput.index(), Map.of());
+        }
+
+        // we assume that all ongoing recoveries will complete
+        for (final var routingNode : routingNodes) {
+            for (final var shardRouting : routingNode) {
+                if (shardRouting.initializing()) {
+                    clusterInfoSimulator.simulate(shardRouting);
+                    routingNodes.startShard(logger, shardRouting, changes, 0L);
+                }
+            }
+        }
+
+        // we are not responsible for allocating unassigned primaries of existing shards, and we're only responsible for allocating
+        // unassigned replicas if the ReplicaShardAllocator gives up, so we must respect these ignored shards
+        final var unassignedPrimaries = new HashSet<ShardId>();
+        final var shardRoutings = new HashMap<ShardId, ShardRoutings>();
+        for (final var primary : new boolean[] { true, false }) {
+            final RoutingNodes.UnassignedShards unassigned = routingNodes.unassigned();
+            for (final var iterator = unassigned.iterator(); iterator.hasNext();) {
+                final var shardRouting = iterator.next();
+                if (shardRouting.primary() == primary) {
+                    var lastAllocatedNodeId = shardRouting.unassignedInfo().getLastAllocatedNodeId();
+                    if (knownNodeIds.contains(lastAllocatedNodeId) || ignoredShards.contains(shardRouting) == false) {
+                        shardRoutings.computeIfAbsent(shardRouting.shardId(), ShardRoutings::new).unassigned().add(shardRouting);
+                    } else {
+                        iterator.removeAndIgnore(UnassignedInfo.AllocationStatus.NO_ATTEMPT, changes);
+                        if (shardRouting.primary()) {
+                            unassignedPrimaries.add(shardRouting.shardId());
+                        }
+                    }
+                }
+            }
+        }
+
+        for (final var assigned : routingNodes.getAssignedShards().entrySet()) {
+            shardRoutings.computeIfAbsent(assigned.getKey(), ShardRoutings::new).assigned().addAll(assigned.getValue());
+        }
+
+        // we can assume that all possible shards will be allocated/relocated to one of their desired locations
+        final var unassignedShardsToInitialize = new HashMap<ShardRouting, LinkedList<String>>();
+        for (final var entry : shardRoutings.entrySet()) {
+            final var shardId = entry.getKey();
+            final var routings = entry.getValue();
+
+            // treesets so that we are consistent about the order of future relocations
+            final var shardsToRelocate = new TreeSet<>(Comparator.comparing(ShardRouting::currentNodeId));
```

Comment:
This is an invariant of `RoutingNodes`, see e.g.

https://github.com/elastic/elasticsearch/blob/30f24d4eea7cb44d78561fd2c88ba477fb5d3d70/server/src/main/java/org/elasticsearch/cluster/routing/RoutingNode.java#L131-L146

Reply:
Thanks. My comment was mostly around using a `TreeSet` that does not adhere to the `Set` contract. I find that quite detailed and it takes me down a path of understanding why and whether it is safe rather than understanding the code. Hence the request to move to a `TreeMap` to make it explicit (and adhere to the contract).

I do think you are right, that this will work as is.

Reply:
Oh yikes I missed that the comparator is used for equality tests. That wasn't my intention here and I see the confusion now. 0e0684f123afb72dd86959bb0dbb00c701ae28e6 addresses this.